db/src/rocksdb.rs

```diff
@@ -1,49 +1,116 @@
 use crate::{Col, DBConfig, DbBatch, Error, KeyValueDB, Result};
-use log::warn;
+use log::{info, warn};
 use rocksdb::{ColumnFamily, Error as RdbError, Options, WriteBatch, DB};
 use std::ops::Range;
 use std::sync::Arc;
 
+// If any data format in database was changed, we have to update this constant manually.
+//      - If the data can be migrated at startup automatically: update "x.y.z1" to "x.y.z2".
+//      - If the data can be migrated manually: update "x.y1.z" to "x.y2.0".
+//      - If the data can not be migrated: update "x1.y.z" to "x2.0.0".
+pub(crate) const VERSION_KEY: &str = "db-version";
+pub(crate) const VERSION_VALUE: &str = "0.1.0";
+
 pub struct RocksDB {
     inner: Arc<DB>,
 }
 
 impl RocksDB {
-    pub fn open(config: &DBConfig, columns: u32) -> Self {
+    pub(crate) fn open_with_check(
+        config: &DBConfig,
+        columns: u32,
+        ver_key: &str,
+        ver_val: &str,
+    ) -> Result<Self> {
         let mut opts = Options::default();
-        opts.create_if_missing(true);
+        opts.create_if_missing(false);
         opts.create_missing_column_families(true);
 
         let cfnames: Vec<_> = (0..columns).map(|c| c.to_string()).collect();
         let cf_options: Vec<&str> = cfnames.iter().map(|n| n as &str).collect();
-        let db = DB::open_cf(&opts, &config.path, &cf_options).unwrap_or_else(|err| {
-            if err.as_ref().starts_with("Corruption:") {
-                warn!("Try repairing the rocksdb since {} ...", err);
+
+        let db = DB::open_cf(&opts, &config.path, &cf_options).or_else(|err| {
+            let err_str = err.as_ref();
+            if err_str.starts_with("Invalid argument:")
```

Comment:
It is not reliable to depend on the error string. It is easier to just check whether the key is present in the database.

Reply:
No, there are two situation:
- (passed) database is not existed
- (failed) database is existed but no version is present

Reply:
It's OK to assume all the guys using the older program will clear their database.